File: frontend/src/services/weatherService.js
const WEATHER_BASE_URL =
  "https://api.open-meteo.com/v1/forecast";


/* =========================================================
   WEATHER RISK CALCULATOR
========================================================= */

function calculateWeatherRisk({
  precipitation = 0,
  rain = 0,
  windSpeed = 0,
}) {

  let risk = 0;


  /* Rain / precipitation */

  if (precipitation >= 20) {
    risk += 45;
  } else if (precipitation >= 10) {
    risk += 30;
  } else if (precipitation >= 5) {
    risk += 20;
  } else if (precipitation > 0) {
    risk += 10;
  }


  /* Rain intensity */

  if (rain >= 20) {
    risk += 25;
  } else if (rain >= 10) {
    risk += 15;
  } else if (rain >= 5) {
    risk += 10;
  }


  /* Wind */

  if (windSpeed >= 60) {
    risk += 30;
  } else if (windSpeed >= 40) {
    risk += 20;
  } else if (windSpeed >= 25) {
    risk += 10;
  }


  return Math.min(
    Math.round(risk),
    100
  );
}


/* =========================================================
   WEATHER CONDITION
========================================================= */

function getWeatherCondition(
  weatherCode
) {

  const code =
    Number(weatherCode);


  if (code === 0) {
    return "Clear";
  }


  if (
    code === 1 ||
    code === 2 ||
    code === 3
  ) {
    return "Cloudy";
  }


  if (
    code === 45 ||
    code === 48
  ) {
    return "Fog";
  }


  if (
    code >= 51 &&
    code <= 57
  ) {
    return "Drizzle";
  }


  if (
    code >= 61 &&
    code <= 67
  ) {
    return "Rain";
  }


  if (
    code >= 71 &&
    code <= 77
  ) {
    return "Snow";
  }


  if (
    code >= 80 &&
    code <= 82
  ) {
    return "Rain Showers";
  }


  if (
    code >= 85 &&
    code <= 86
  ) {
    return "Snow Showers";
  }


  if (
    code >= 95 &&
    code <= 99
  ) {
    return "Thunderstorm";
  }


  return "Unknown";
}


/* =========================================================
   GET WEATHER
========================================================= */

export async function getWeather(
  latitude,
  longitude
) {

  const lat =
    Number(latitude);

  const lng =
    Number(longitude);


  /* =======================================================
     VALIDATION
  ======================================================= */

  if (
    !Number.isFinite(lat) ||
    !Number.isFinite(lng)
  ) {

    throw new Error(
      "Invalid weather coordinates."
    );

  }


  if (
    lat < -90 ||
    lat > 90
  ) {

    throw new Error(
      "Invalid latitude."
    );

  }


  if (
    lng < -180 ||
    lng > 180
  ) {

    throw new Error(
      "Invalid longitude."
    );

  }


  /* =======================================================
     API URL
  ======================================================= */

  const url =
    `${WEATHER_BASE_URL}` +
    `?latitude=${lat}` +
    `&longitude=${lng}` +
    `&current=temperature_2m,relative_humidity_2m,precipitation,rain,weather_code,wind_speed_10m` +
    `&timezone=auto`;


  /* =======================================================
     REQUEST
  ======================================================= */

  const response =
    await fetch(url);


  if (!response.ok) {

    throw new Error(
      `Weather service failed (${response.status}).`
    );

  }


  const data =
    await response.json();


  if (
    !data?.current
  ) {

    throw new Error(
      "Weather data unavailable."
    );

  }


  /* =======================================================
     CURRENT DATA
  ======================================================= */

  const current =
    data.current;


  const temperature =
    Number(
      current.temperature_2m
    ) || 0;


  const humidity =
    Number(
      current.relative_humidity_2m
    ) || 0;


  const precipitation =
    Number(
      current.precipitation
    ) || 0;


  const rain =
    Number(
      current.rain
    ) || 0;


  const windSpeed =
    Number(
      current.wind_speed_10m
    ) || 0;


  const weatherCode =
    Number(
      current.weather_code
    );


  /* =======================================================
     CONDITION
  ======================================================= */

  const condition =
    getWeatherCondition(
      weatherCode
    );


  /* =======================================================
     WEATHER RISK
  ======================================================= */

  let weatherRisk =
    calculateWeatherRisk({

      precipitation,

      rain,

      windSpeed,

    });


  /* =======================================================
     EXTRA THUNDERSTORM RISK
  ======================================================= */

  if (
    weatherCode >= 95
  ) {

    weatherRisk =
      Math.min(
        weatherRisk + 30,
        100
      );

  }


  /* =======================================================
     RETURN STANDARD FORMAT
  ======================================================= */

  return {

    latitude: lat,

    longitude: lng,

    condition,

    temperature,

    humidity,

    precipitation,

    rain,

    windSpeed,

    weatherCode,

    weatherRisk,

    source: "Open-Meteo",

    updatedAt:
      new Date().toISOString(),

  };

}

File: backend/app/services/weather_service.py
import httpx

from app.schemas.weather import WeatherResponse

WEATHER_CODES = {
    0: "Clear sky",
    1: "Mainly clear",
    2: "Partly cloudy",
    3: "Overcast",
    45: "Fog",
    48: "Depositing rime fog",
    51: "Light drizzle",
    53: "Moderate drizzle",
    55: "Dense drizzle",
    61: "Slight rain",
    63: "Moderate rain",
    65: "Heavy rain",
    71: "Slight snow",
    73: "Moderate snow",
    75: "Heavy snow",
    80: "Slight rain showers",
    81: "Moderate rain showers",
    82: "Violent rain showers",
    95: "Thunderstorm",
    96: "Thunderstorm with light hail",
    99: "Thunderstorm with heavy hail"
}

async def get_weather_for_location(lat: float, lon: float) -> WeatherResponse:
    url = f"https://api.open-meteo.com/v1/forecast?latitude={lat}&longitude={lon}&current=temperature_2m,is_day,precipitation,weather_code,wind_speed_10m"
    
    async with httpx.AsyncClient() as client:
        try:
            response = await client.get(url, timeout=5.0)
            response.raise_for_status()
            data = response.json()
            
            current = data.get("current", {})
            w_code = current.get("weather_code", 0)
            
            return WeatherResponse(
                temperature_celsius=current.get("temperature_2m", 0.0),
                weather_code=w_code,
                weather_description=WEATHER_CODES.get(w_code, "Unknown"),
                wind_speed_kmh=current.get("wind_speed_10m", 0.0),
                precipitation_mm=current.get("precipitation", 0.0),
                is_day=bool(current.get("is_day", 1))
            )
        except Exception as e:
            # Return a fallback if API fails
            print(f"Weather API failed: {e}")
            return WeatherResponse(
                temperature_celsius=25.0,
                weather_code=0,
                weather_description="Clear sky",
                wind_speed_kmh=5.0,
                precipitation_mm=0.0,
                is_day=True
            )
